File: backend/app/schemas/relationship.py
"""Request/response schemas for the Asset Relationships API."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.relationship import DESCRIPTION_MAX_LENGTH, RelationshipType

MAX_PAGE_SIZE = 100
DEFAULT_PAGE_SIZE = 50


class AssetSummaryRead(BaseModel):
    """Safe, minimal Asset projection embedded in a relationship response - never
    the full :class:`~app.schemas.asset.AssetRead` (avoids over-fetching and
    keeps the relationship payload small)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    hostname: str | None
    asset_type: str
    environment: str
    criticality: str
    status: str
    is_active: bool


class RelationshipCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    source_asset_id: uuid.UUID
    target_asset_id: uuid.UUID
    relationship_type: RelationshipType
    description: str | None = Field(default=None, max_length=DESCRIPTION_MAX_LENGTH)

    @model_validator(mode="after")
    def _no_self_link(self) -> RelationshipCreate:
        if self.source_asset_id == self.target_asset_id:
            raise ValueError("source_asset_id and target_asset_id must be different")
        return self

    @model_validator(mode="after")
    def _clean_description(self) -> RelationshipCreate:
        if self.description is not None:
            cleaned = self.description.strip()
            object.__setattr__(self, "description", cleaned or None)
        return self


class RelationshipUpdate(BaseModel):
    """Partial update. Only ``relationship_type`` / ``description`` are
    editable - source/target are immutable through PATCH (§13): delete the
    edge and create a new one if it needs to point elsewhere."""

    model_config = ConfigDict(extra="forbid")

    relationship_type: RelationshipType | None = None
    description: str | None = Field(default=None, max_length=DESCRIPTION_MAX_LENGTH)
    #: True once the client actually sent ``description`` (distinguishes "clear
    #: it" from "leave it alone") - mirrors the ``exclude_unset`` pattern used
    #: elsewhere in the codebase via ``model_dump(exclude_unset=True)``.


class RelationshipRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source_asset_id: uuid.UUID
    target_asset_id: uuid.UUID
    relationship_type: RelationshipType
    description: str | None
    created_at: datetime
    updated_at: datetime


class RelationshipDetail(RelationshipRead):
    """A relationship with both endpoint Assets embedded - what the list /
    grouped / detail endpoints actually return, so the frontend never has to
    issue a follow-up fetch per edge."""

    source: AssetSummaryRead
    target: AssetSummaryRead


class RelationshipPage(BaseModel):
    items: list[RelationshipDetail]
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=MAX_PAGE_SIZE)
    total: int = Field(ge=0)
    total_pages: int = Field(ge=0)


class RelationshipCounts(BaseModel):
    outgoing: int = Field(ge=0)
    incoming: int = Field(ge=0)
    total: int = Field(ge=0)


class AssetRelationshipsGrouped(BaseModel):
    """The Dependencias-tab shape: this Asset's own edges grouped by direction."""

    outgoing: list[RelationshipDetail]
    incoming: list[RelationshipDetail]
    counts: RelationshipCounts


class RelationshipTypeInfo(BaseModel):
    """One entry of the relationship-type catalog, for frontend labels /
    the Add-relationship type select - a single source of truth."""

    code: str
    label: str
    inverse_label: str
    description: str
    category: str
    directed: bool = True
    propagates_impact: bool


class RelationshipTypeCatalog(BaseModel):
    types: list[RelationshipTypeInfo]


class RelationshipSummary(BaseModel):
    """Restrained global stats for the Dependencias module header - not a
    dashboard. ``assets_without_relationships`` is included since it is a
    single cheap extra count, not a wider aggregation."""

    total: int = Field(ge=0)
    connected_assets: int = Field(ge=0)
    relationship_types: int = Field(ge=0)
    assets_without_relationships: int = Field(ge=0)


class MessageResponse(BaseModel):
    detail: str

File: frontend/src/types/relationship.ts
/**
 * Asset relationship domain types (Asset Relationships & Topology milestone).
 * PostgreSQL is canonical; these mirror `app/schemas/relationship.py` exactly.
 */

export const RELATIONSHIP_TYPES = [
  "depends_on",
  "hosts",
  "connects_to",
  "uses",
  "provides_service_to",
  "member_of",
] as const;
export type RelationshipType = (typeof RELATIONSHIP_TYPES)[number];

export interface AssetSummary {
  id: string;
  name: string;
  hostname: string | null;
  asset_type: string;
  environment: string;
  criticality: string;
  status: string;
  is_active: boolean;
}

export interface RelationshipSummary {
  total: number;
  connected_assets: number;
  relationship_types: number;
  assets_without_relationships: number;
}

export interface RelationshipDetail {
  id: string;
  source_asset_id: string;
  target_asset_id: string;
  relationship_type: RelationshipType;
  description: string | null;
  created_at: string;
  updated_at: string;
  source: AssetSummary;
  target: AssetSummary;
}

export interface RelationshipPage {
  items: RelationshipDetail[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface RelationshipCounts {
  outgoing: number;
  incoming: number;
  total: number;
}

export interface AssetRelationshipsGrouped {
  outgoing: RelationshipDetail[];
  incoming: RelationshipDetail[];
  counts: RelationshipCounts;
}

export interface RelationshipTypeInfo {
  code: RelationshipType;
  label: string;
  inverse_label: string;
  description: string;
  category: string;
  directed: boolean;
  propagates_impact: boolean;
}

export interface RelationshipTypeCatalog {
  types: RelationshipTypeInfo[];
}

export interface RelationshipCreateInput {
  source_asset_id: string;
  target_asset_id: string;
  relationship_type: RelationshipType;
  description?: string | null;
}

export interface RelationshipUpdateInput {
  relationship_type?: RelationshipType;
  description?: string | null;
}

// --- runtime guards ---------------------------------------------------------

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function isAssetSummary(v: unknown): v is AssetSummary {
  return (
    isRecord(v) &&
    typeof v.id === "string" &&
    typeof v.name === "string" &&
    (v.hostname === null || v.hostname === undefined || typeof v.hostname === "string") &&
    typeof v.asset_type === "string" &&
    typeof v.environment === "string" &&
    typeof v.criticality === "string" &&
    typeof v.status === "string" &&
    typeof v.is_active === "boolean"
  );
}

export function isRelationshipSummary(v: unknown): v is RelationshipSummary {
  return (
    isRecord(v) &&
    typeof v.total === "number" &&
    typeof v.connected_assets === "number" &&
    typeof v.relationship_types === "number" &&
    typeof v.assets_without_relationships === "number"
  );
}

export function isRelationshipDetail(v: unknown): v is RelationshipDetail {
  return (
    isRecord(v) &&
    typeof v.id === "string" &&
    typeof v.source_asset_id === "string" &&
    typeof v.target_asset_id === "string" &&
    typeof v.relationship_type === "string" &&
    (v.description === null || typeof v.description === "string") &&
    typeof v.created_at === "string" &&
    typeof v.updated_at === "string" &&
    isAssetSummary(v.source) &&
    isAssetSummary(v.target)
  );
}

export function isRelationshipPage(v: unknown): v is RelationshipPage {
  return (
    isRecord(v) &&
    Array.isArray(v.items) &&
    v.items.every(isRelationshipDetail) &&
    typeof v.page === "number" &&
    typeof v.total === "number" &&
    typeof v.total_pages === "number"
  );
}

export function isAssetRelationshipsGrouped(v: unknown): v is AssetRelationshipsGrouped {
  return (
    isRecord(v) &&
    Array.isArray(v.outgoing) &&
    v.outgoing.every(isRelationshipDetail) &&
    Array.isArray(v.incoming) &&
    v.incoming.every(isRelationshipDetail) &&
    isRecord(v.counts) &&
    typeof v.counts.outgoing === "number" &&
    typeof v.counts.incoming === "number" &&
    typeof v.counts.total === "number"
  );
}

export function isRelationshipTypeCatalog(v: unknown): v is RelationshipTypeCatalog {
  return (
    isRecord(v) &&
    Array.isArray(v.types) &&
    v.types.every(
      (t) =>
        isRecord(t) &&
        typeof t.code === "string" &&
        typeof t.label === "string" &&
        typeof t.inverse_label === "string" &&
        typeof t.propagates_impact === "boolean",
    )
  );
}
